### DISCUSSION TITLE
Receiving via AMQP

### DISCUSSION THREAD

**Discussion by snv** [smallrye/smallrye-reactive-messaging#2620] [Q&A]
Hi,

i try to receive a simple DTO via amqp, but keep running into 
> ERROR [io.smallrye.reactive.messaging.amqp] (vert.x-eventloop-thread-0) SRMSG16225: Failure reported for channel `amqpGreetingsIn`, closing client: io.smallrye.reactive.messaging.providers.ProcessingException: SRMSG00103: Exception thrown when calling the method org.eclipse.jakarta.hello.MessagingService#consumeAmqp
[..]
Caused by: java.lang.IllegalArgumentException: argument type mismatch
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:107)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.smallrye.reactive.messaging@4.11.0//io.smallrye.reactive.messaging.providers.AbstractMediator.lambda$initialize$9(AbstractMediator.java:133)

According to https://smallrye.io/smallrye-reactive-messaging/smallrye-reactive-messaging/3.3/amqp/amqp.html#_deserialization i should need to receive a io.vertx.core.json.JsonObject like
```  
  @Incoming("amqpGreetingsIn")
  public void consumeAmqp(JsonObject greeting){
    LOG.info("Greetings for " + greeting.mapTo(Hello.class).getHello() + " received via AMQP");
  }
```
but this refuses to work with a message i just sent via 
```	
        @Inject
	@Channel("amqpGreetingsOut")
	Emitter<Hello> amqpHelloEmitter;
```

To send i can get by with just the microprofile api, but for receiving jsons, do i really have to add vertx implementation as a dependency?
And why am i getting these exceptions which even shutdown wildfly, what am i doing wrong there?

**Comment by cescoffier** [comment]
Yes, on the receiving side, it needs to be a JSONObject. You can write a converter, if you want, to map it to your own DTO automatically (`object.mapTo(DTO.class)`)
